### PR TITLE
CP-11040: Support {attach_from,generate}_config in EXTSR and FileSR

### PR DESCRIPTION
**Pull request by djs55**
These functions allow VDIs to be attached 'statically' i.e. at host
boot time. They allow VDIs to be used for the Xapi database redo log
which prevents database updates being lost.

Signed-off-by: David Scott <dave.scott@citrix.com>

### CODE REVIEW

**Comment by djs55** [comment]
Problem with travis configuration?
```
17.30s$ pip install -r requirements.txt --use-mirrors
You are using pip version 6.0.7, however version 6.0.8 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
DEPRECATION: --use-mirrors has been deprecated and will be removed in the future. Explicit uses of --index-url and/or --extra-index-url is suggested.
...
Successfully installed PyYAML-3.11 astroid-1.3.4 coverage-3.7.1 coveralls-0.5 docopt-0.6.2 logilab-common-0.63.2 pylint-1.2.0 requests-2.5.1 six-1.9.0 unittest2-0.8.0 xenapi-1.2
0.27s$ pylint --version
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.6.9/bin/pylint", line 9, in <module>
    load_entry_point('pylint==1.2.0', 'console_scripts', 'pylint')()
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/pylint/__init__.py", line 20, in run_pylint
    from pylint.lint import Run
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/pylint/lint.py", line 30, in <module>
    from pylint.checkers import utils
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/pylint/checkers/__init__.py", line 45, in <module>
    from astroid.utils import ASTWalker
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/astroid/__init__.py", line 54, in <module>
    from astroid.nodes import *
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/astroid/nodes.py", line 48, in <module>
    from astroid.scoped_nodes import Module, GenExpr, Lambda, DictComp, \
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/astroid/scoped_nodes.py", line 48, in <module>
    from astroid.manager import AstroidManager
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/astroid/manager.py", line 36, in <module>
    from astroid import modutils
  File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/astroid/modutils.py", line 69
    get_python_lib(standard_lib=True, prefix=sys.prefix),
                                                        ^
SyntaxError: invalid syntax
```

**Comment by germanop** [comment]
@djs55, no major concerns about this patch.
I assume you want this feature to be available just for FileSR and and EXTSR, so when we will rearrange the code to make it more "inheritance friendly" you do not want this properties to be available for all the file based SRs (e.g. NFS).

Is that correct?

**Comment by germanop** [comment]
OK, maybe I said it was ok too early :)

See my comments inline

File: drivers/FileSR.py [hunk 1]
```diff
@@ -913,6 +917,31 @@ def extractUuid(path):
         return uuid
     extractUuid = staticmethod(extractUuid)
 
+    def generate_config(self, sr_uuid, vdi_uuid):
```
Comment (germanop) [inline drivers/FileSR.py]: Function description would be helpful

File: drivers/FileSR.py [hunk 2]
```diff
@@ -913,6 +917,31 @@ def extractUuid(path):
...
+        resp['vdi_uuid'] = vdi_uuid
+        resp['command'] = 'vdi_attach_from_config'
+        # Return the 'config' encoded within a normal XMLRPC response so that
+        # we can use the regular response/error parsing code.
+        config = xmlrpclib.dumps(tuple([resp]), "vdi_attach_from_config")
+        return xmlrpclib.dumps((config,), "", True)
+
+    def attach_from_config(self, sr_uuid, vdi_uuid):
+        """Used for HA State-file only. Will not just attach the VDI but
+        also start a tapdisk on the file"""
```
Comment (germanop) [inline drivers/FileSR.py]: For multiline docstrings the closing quotes go and a line of their own

File: drivers/FileSR.py [hunk 3]
```diff
@@ -913,6 +917,31 @@ def extractUuid(path):
...
+        util.SMlog("FileVDI.attach_from_config")
+        try:
+           if not util.pathexists(self.sr.path):
+                self.sr.attach(sr_uuid)
+        except:
+            util.logException("FileVDI.attach_from_config")
+            raise xs_errors.XenError('SRUnavailable', \
```
Comment (germanop) [inline drivers/FileSR.py]: No backslashes in parentheses

File: drivers/FileSR.py [hunk 4]
```diff
@@ -913,6 +917,31 @@ def extractUuid(path):
...
+                        opterr='Unable to attach from config')
```
Comment (germanop) [inline drivers/FileSR.py]: Sorry to nitpick but even if it does not seem so we are striving to adhere tp PEP-08
https://www.python.org/dev/peps/pep-0008/#indentation

You can either do this
```python
raise xs_errors.XenError(
    'SRUnavailable',
    opterr='Unable to attach from config'
)
```
this
```python
raise xs_errors.XenError(
    'SRUnavailable',
    opterr='Unable to attach from config')
```
or this
```python
raise xs_errors.XenError('SRUnavailable',
                         opterr='Unable to attach from config')
```

File: drivers/FileSR.py [hunk 5]
```diff
@@ -913,6 +917,31 @@ def extractUuid(path):
...
+        # we can use the regular response/error parsing code.
+        config = xmlrpclib.dumps(tuple([resp]), "vdi_attach_from_config")
+        return xmlrpclib.dumps((config,), "", True)
+
+    def attach_from_config(self, sr_uuid, vdi_uuid):
+        """Used for HA State-file only. Will not just attach the VDI but
+        also start a tapdisk on the file"""
+        util.SMlog("FileVDI.attach_from_config")
+        try:
+           if not util.pathexists(self.sr.path):
```
Comment (germanop) [inline drivers/FileSR.py]: if is indent 3 lines, just add one space here

File: drivers/FileSR.py [hunk 6]
```diff
@@ -913,6 +917,31 @@ def extractUuid(path):
         return uuid
     extractUuid = staticmethod(extractUuid)
 
+    def generate_config(self, sr_uuid, vdi_uuid):
+        util.SMlog("FileVDI.generate_config")
+        if not util.pathexists(self.path):
+                raise xs_errors.XenError('VDIUnavailable')
```
Comment (germanop) [inline drivers/FileSR.py]: too much indentation

**Comment by djs55** [comment]
@germanop: I've tried to address your formatting comments in the new commits. Let me know if I should squash them.

Regarding other SRs: I think NFS already exposes this capability. I suspect that, in future, we'll want to expose the capability in FileSR and then 'inherit' it somehow in all the child classes. I'm not sure what the best way of doing that is.

**Comment by germanop** [comment]
- [ ] Multiline docstring closing
- [ ] generate_config description
- [ ] No backslashes in attach_from_config except block
- [ ] Same block, better functio indentation
- [ ] if indentation in attach_from_config
- [ ] In generate_config too much indentation after if

**Comment by germanop** [comment]
@djs55, do not mind about squashing, I will do that as long it is clear which commit message
has to be retained. It is clear in this case.
In general you can rewrite the pull request because I have a copy of your previous one :) (it would be better if github did that for me..) and comments on the file tab are not lost one commit rewrite.

About future improvements, I did not notice it was already in the NFS code, so this is going to be true
for every file-based SR.
In this case I addressed a similar issue in the past and you can find it in #237 

We will go back to this once my pull request will be in. Technically it has been approved but it was
not urgent.
Give it a look if you want and let me know.

We can do the same for your code (apart fro the class constructor bit that is a one-off)

**Comment by germanop** [comment]
Pushed with commit cfe77a3

**Pull request closed without merging**